### ISSUE TITLE
Authentication against custom schemes does not trigger custom SignInManager.SignInAsync

### ISSUE THREAD

**Issue by hempels** [aspnet/AspNetCore#14984]
We use Identity (via services.AddIdentity<>..) and register a custom SignInManager. This all works correctly with Google auth (.AddAuthentication().AddGoogle())...

Separately we have a custom AuthenticationHandler overriding HandleAuthenticateAsync, which upon success, returns AuthenticateResult.Success with a populated AuthenticationTicket containing an appropriate ClaimsPrincipal, etc. This is registered via .AddAuthentication().AddScheme, and seems to work well on the surface. We can Authorize individual actions with the scheme, users are authenticated, and an Identity cookie is created.

However, we realized that when the custom AuthenticationHandler is in play, **our custom SignInManager is never called**. As a result, some key functionality is lost.

Is there something we're missing when using AddScheme or in the implementation of our AuthenticationHandler that is needed in order to have it also utilize the SignInManager?

Using ASP.NET Core 2.1 compatibility version on 2.2.7 runtime.

**Comment by hempels** [comment]
A bit more color.

From within HandleAutenticateAsync I can retreive the registered SignInManager and call SignInAsync on that direction, which works as desired.

It is only if I try to call Context.SignInAsync(IdentityConstants.ApplicationScheme, principal, ...) that the custom SignInManager is ignored.

So I can further narrow my question. Is there something we need to do in order to get the HttpContext.SignInAsync extension method to make use of our registered SignInManager?

**Comment by blowdart** [comment]
I believe SignInManager will only kick in when identity & cookie authentication is triggered, which would happen with Google because that will flow through the identity pieces.

@HaoK to confirm, because I'm only on my second coffee.

**Comment by Tratcher** [comment]
Google and other remote auth providers call Context.SignInAsync with the ExternalScheme, not the ApplicationScheme. There's then a second phase where Identity exchanged the external cookie for the app cookie via SignInManager.
https://github.com/aspnet/AspNetCore/blob/21c9e2cc954c10719878839cd3f766aca5f57b34/src/Security/Authentication/Core/src/RemoteAuthenticationHandler.cs#L140
https://github.com/aspnet/AspNetCore/blob/62351067ff4c1401556725b401478e648b66acdc/src/Security/Authentication/Core/src/AuthenticationServiceCollectionExtensions.cs#L61

**Comment by hempels** [comment]
What I'm having trouble with is figuring out how to implement that exchange from within our custom auth handler. I appreciate that the RemoteAuthenticationHandler does it, but digging through that code is not a simple task and I have not been able to figure out where that magic actually happens.

The basic setup is that we have a single action method set via attribute to authorize against a custom scheme. The custom scheme's auth handler is meant to result in a normal ApplicationScheme authentication and return a success ticket so the action method will run.

On the surface it appears to work as is, but once I drill in further it becomes apparent that none of the authentication properties or claims are being carried through into the ApplicationScheme auth.

**Comment by hempels** [comment]
I think I worked out the issue. In the method that's utilizing the custom scheme, it was calling RefreshSignInAsync. Although I still don't quite understand why, at that point any calls to Context.AuthenticateAsync() will not return the populated Identity Scheme data. However, once the identity cookie is set, calls from other sites do return the correct data.

I had to replace that call to RefreshSignInAsync within that method with a new call to SignInAsync and explicitly pass the properties that need to be preserved.

It still feels a little bit wrong, like some part of the middleware is not realizing that an identity auth has taken place. But it's working for us.

**Comment by Tratcher** [comment]
The exchange doesn't happen in the auth handlers, it happens in an MVC controller (in Identity.UI?).

**Comment by HaoK** [comment]
Yeah the callback happens here https://github.com/aspnet/AspNetCore/blob/master/src/Identity/UI/src/Areas/Identity/Pages/V4/Account/ExternalLogin.cshtml.cs#L132

**Comment by hempels** [comment]
Right. I realized that later, which is why I deleted the comment you were probably responding to.

I still find it curious that in the method that's triggering our custom auth handler, any calls to Context.AuthenticateAsync do not return the information. Only after the cookie is set and a different action is executed (utilizing the identity auth) does that function do what I expect.

If I have more time I'll try this with the templates, but I suspect in that ExternalLogin.cshtml.cs code, if you were to execute Context.AuthenticateAsync() after the call to _signInManager.SignInAsync ([here](https://github.com/aspnet/AspNetCore/blob/master/src/Identity/UI/src/Areas/Identity/Pages/V4/Account/ExternalLogin.cshtml.cs#L206)) that it would not return the expected information. Unless something has changed since 2.2.
